File: thinking-categorically/applicative-functor/reader-applicative/async-reader-applicatve.cs
using System;

public class Reader<R, A>
{
    public Func<R, A> run { get; }

    public Reader(Func<R, A> run)
    {
        this.run = run;
    }

    // Run the computation
    public A Run(R env) => run(env);

    // Functor: Select (map)
    public Reader<R, B> Select<B>(Func<A, B> f)
    {
        return new Reader<R, B>(r => f(run(r)));
    }

    // Monad: SelectMany (bind / flatMap)
    public Reader<R, B> SelectMany<B>(Func<A, Reader<R, B>> f)
    {
        return new Reader<R, B>(r =>
        {
            var a = run(r);
            return f(a).run(r);
        });
    }

    // For LINQ query with projection after SelectMany
    public Reader<R, C> SelectMany<B, C>(Func<A, Reader<R, B>> f, Func<A, B, C> projector)
    {
        return new Reader<R, C>(r =>
        {
            var a = run(r);
            var b = f(a).run(r);
            return projector(a, b);
        });
    }

    // Applicative pure
    public static Reader<R, A> Pure(A value)
    {
        return new Reader<R, A>(_ => value);
    }
}

class Env
{
    public int Multiplier { get; set; }
}

class Program
{
    static void Main()
    {
        var readerA = new Reader<Env, int>(env => 5 + env.Multiplier);
        var readerB = new Reader<Env, int>(env => 10 * env.Multiplier);

        // Using LINQ query syntax
        var combined = from a in readerA
                       from b in readerB
                       select a + b;

        var result = combined.Run(new Env { Multiplier = 3 }); // (5+3) + (10*3) = 8 + 30 = 38
        Console.WriteLine(result);  // Output: 38
    }
}

File: thinking-categorically/applicative-functor/reader-applicative/async-reader-applicatve.ts
// Async Reader: environment R -> Promise<A>
type ReaderAsync<R, A> = (r: R) => Promise<A>;

class ReaderT<R, A> {
  constructor(public readonly run: ReaderAsync<R, A>) {}

  // Functor: map
  map<B>(f: (a: A) => B): ReaderT<R, B> {
    return new ReaderT(async (r: R) => {
      const a = await this.run(r);
      return f(a);
    });
  }

  // Applicative: ap (parallel async version)
  ap<B>(rf: ReaderT<R, (a: A) => B>): ReaderT<R, B> {
    return new ReaderT(async (r: R) => {
      // Run both in parallel
      const [func, val] = await Promise.all([rf.run(r), this.run(r)]);
      return func(val);
    });
  }

  // Applicative: pure
  static of<R, A>(value: A): ReaderT<R, A> {
    return new ReaderT(() => Promise.resolve(value));
  }
}

// Environment type
type Env = { multiplier: number };

// Simulated async computation that adds to the multiplier
const readerA = new ReaderT<Env, number>(async (env) => {
  await delay(100); // simulate latency
  return 5 + env.multiplier;
});

// Simulated async computation that produces a function
const readerF = new ReaderT<Env, (x: number) => number>(async (env) => {
  await delay(100); // simulate latency
  return (x: number) => x * env.multiplier;
});

// Run both computations in parallel and apply result
const resultReader = readerA.ap(readerF);

// Run with environment
(async () => {
  const result = await resultReader.run({ multiplier: 3 });
  console.log(result); // (5 + 3) * 3 = 8 * 3 = 24
})();

// Helper delay function
function delay(ms: number): Promise<void> {
  return new Promise(res => setTimeout(res, ms));
}
